In [1]:
import pandas as pd

In [2]:
# Завантажуємо основні датасети
df_p = pd.read_csv ('/Users/irynachystiskova/Downloads/stack-overflow-developer-survey-2025/survey_results_public.csv', 
                    encoding = 'utf-8',
                   low_memory=False)

df_s = pd.read_csv ('/Users/irynachystiskova/Downloads/stack-overflow-developer-survey-2025/survey_results_schema.csv', 
                    encoding = 'utf-8',
                   low_memory=False)


In [3]:
# Перевіряємо, що датасети коректно завантажилися (перші рядки head та розмір shape)

In [4]:
df_p.head()

,ResponseId,MainBranch,Age,EdLevel,Employment,EmploymentAddl,WorkExp,LearnCodeChoose,LearnCode,LearnCodeAI,...,AIAgentOrchestration,AIAgentOrchWrite,AIAgentObserveSecure,AIAgentObsWrite,AIAgentExternal,AIAgentExtWrite,AIHuman,AIOpen,ConvertedCompYearly,JobSat
0,1,I am a developer by profession,25-34 years old,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)",Employed,"Caring for dependents (children, elderly, etc.)",8.0,"Yes, I am not new to coding but am learning ne...",Online Courses or Certification (includes all ...,"Yes, I learned how to use AI-enabled tools for...",...,Vertex AI,NaN,NaN,NaN,ChatGPT,NaN,When I don’t trust AI’s answers,"Troubleshooting, profiling, debugging",61256.0,10.0
1,2,I am a developer by profession,25-34 years old,"Associate degree (A.A., A.S., etc.)",Employed,NaN,2.0,"Yes, I am not new to coding but am learning ne...",Online Courses or Certification (includes all ...,"Yes, I learned how to use AI-enabled tools for...",...,NaN,NaN,NaN,NaN,NaN,NaN,When I don’t trust AI’s answers;When I want to...,All skills. AI is a flop.,104413.0,9.0
2,3,I am a developer by profession,35-44 years old,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)","Independent contractor, freelancer, or self-em...",None of the above,10.0,"Yes, I am not new to coding but am learning ne...",Online Courses or Certification (includes all ...,"Yes, I learned how to use AI-enabled tools for...",...,NaN,NaN,NaN,NaN,ChatGPT;Claude Code;GitHub Copilot;Google Gemini,NaN,When I don’t trust AI’s answers;When I want to...,"Understand how things actually work, problem s...",53061.0,8.0
3,4,I am a developer by profession,35-44 years old,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)",Employed,None of the above,4.0,"Yes, I am not new to coding but am learning ne...","Other online resources (e.g. standard search, ...","Yes, I learned how to use AI-enabled tools for...",...,NaN,NaN,NaN,NaN,ChatGPT;Claude Code,NaN,When I don’t trust AI’s answers;When I want to...,NaN,36197.0,6.0
4,5,I am a developer by profession,35-44 years old,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)","Independent contractor, freelancer, or self-em...","Caring for dependents (children, elderly, etc.)",21.0,"No, I am not new to coding and did not learn n...",NaN,"Yes, I learned how to use AI-enabled tools for...",...,NaN,NaN,NaN,NaN,NaN,NaN,When I don’t trust AI’s answers,"critical thinking, the skill to define the tas...",60000.0,7.0


In [5]:
df_s.head()

,qid,qname,question,type,sub,sq_id
0,QID18,TechEndorse_1,What attracts you to a technology or causes yo...,RO,AI integration or AI Agent capabilities,1.0
1,QID18,TechEndorse_2,What attracts you to a technology or causes yo...,RO,Easy-to-use API,2.0
2,QID18,TechEndorse_3,What attracts you to a technology or causes yo...,RO,Robust and complete API,3.0
3,QID18,TechEndorse_4,What attracts you to a technology or causes yo...,RO,Customizable and manageable codebase,4.0
4,QID18,TechEndorse_5,What attracts you to a technology or causes yo...,RO,Reputation for quality,5.0


In [6]:
print (df_p.shape)

(49191, 172)


In [7]:
print (df_s.shape)

(139, 6)


Завдання 1. Підрахунок загальної кількості респондентів 
-
- Інструмент, який потрібен для виконання цього завдання: Pandas (методи для роботи з DataFrame)

- Опис завдання: Тобі потрібно визначити загальну кількість респондентів, які взяли участь в опитуванні Stack Overflow.
Це базова метрика, яка дасть тобі уявлення про обсяг датасету.


In [8]:
# Перевіряємо, що кожен респондент має унікальний ResponseId
df_p['ResponseId'].is_unique

True

In [9]:
# Підрахунок загальної кількості респондентів
total_respondents = df_p.shape[0]
total_respondents

49191

ВИСНОВОК:

У опитуванні Stack Overflow взяли участь 49 191 респондент, що дає уявлення про масштаб датасету та дозволяє проводити подальший аналіз.

Завдання 2. Аналіз повноти відповідей респондентів
-
- Інструмент, який потрібен для виконання цього завдання: Pandas (методи для роботи з пропущеними значеннями, множинами)

- Опис завдання: Визнач, скільки респондентів відповіли на всі запитання опитування.

- Для цього тобі знадобиться:
> - Отримати список питань з файлу survey_results_schema (поле "qname")
> - Перевірити наявність відповідей на ці питання у кожного респондента
> - Підрахувати кількість респондентів без пропущених значень

In [10]:
# Кількість унікальних питань з survey_results_schema
schema_question = set(df_s['qname'])
len(schema_question)

139

In [11]:
# Кількість питань df_s, які перетинаються зі стовпцями df_p
common_cols = list(schema_question.intersection(df_p.columns))
len(common_cols)

126

In [12]:
# Підраховуємо кількість респондентів без жодного пропущеного значення
complete_respondents = df_p[common_cols].dropna()
complete_respondents.shape[0]

0

ВИСНОВОК: 

Аналіз повноти відповідей показав, що жоден респондент не відповів на всі питання опитування. Це пояснюється наявністю умовних та спеціалізованих запитань, які не передбачають обов’язкової відповіді для всіх учасників.

Завдання 3. Статистичний аналіз досвіду респондентів
- 
- Інструмент, який потрібен для виконання цього завдання: Pandas (описова статистика, методи центральної тенденції)

- Опис завдання: Обчисли міри центральної тенденції для поля WorkExp (досвід роботи респондентів).

- Тобі потрібно знайти:
> - Середнє значення (mean)
> - Медіану (median)
> - Моду (mode)

In [13]:
# Вибираємо колонку з досвідом роботи
work_experience = df_p['WorkExp']

# Перевіримо тип даних
work_experience.dtype

dtype('float64')

In [14]:
df_p['WorkExp'].info()
#У колонці WorkExp присутні пропущені значення, однак pandas автоматично виключає їх з розрахунків статистичних показників.

<class 'pandas.core.series.Series'>
RangeIndex: 49191 entries, 0 to 49190
Series name: WorkExp
Non-Null Count  Dtype  
--------------  -----  
42893 non-null  float64
dtypes: float64(1)
memory usage: 384.4 KB


In [15]:
# Створюємо DataFrame з результатами
workexp_stats = pd.DataFrame ({
    'metric' : ['mean', 'median', 'mode'],
    'value' : [
        df_p ['WorkExp'].mean().round(),
        df_p ['WorkExp'].median(),
        df_p ['WorkExp'].mode().iloc[0]
    ]
})
workexp_stats

,metric,value
0,mean,13.0
1,median,10.0
2,mode,10.0


ВИСНОВОК: 

Середній досвід роботи респондентів становить близько 13 років, медіанний - 10 років. Це означає, що половина респондентів має досвід до 10 років. Найпоширеніше значення також дорівнює 10 рокам, що свідчить про концентрацію відповідей навколо цього рівня досвіду. Те, що середнє значення вище за медіану, може вказувати на наявність респондентів з дуже великим досвідом.

Завдання 4. Аналіз віддаленої роботи
-
- Інструмент, який потрібен для виконання цього завдання: Pandas (фільтрація даних, підрахунок значень)

- Опис завдання: Визнач кількість респондентів, які працюють віддалено (remote).

- Тобі потрібно знайти відповідне поле в датасеті та підрахувати кількість працівників, які обрали цей формат роботи.

In [16]:
# Дивимось ВСІ питання, пов’язані з роботою, з метою знайти відповідну колонку в df_p
df_s[df_s['question'].str.contains('work', case=False, na=False)][['qname', 'question']]

,qname,question
54,WorkExp,How many years of professional work experience...
57,LearnCodeAI,Did you spend time in the last year learning A...
61,OrgSize,Approximately how many people are employed by ...
63,RemoteWork,Which best describes your current work situation?
65,TechEndorseIntro,When you&rsquo;re in a position to <strong>inf...
66,Industry,What industry do you work for in?
70,ToolCountWork,"For your <b>primary work role</b> (student, de..."
71,ToolCountPersonal,<b></b>For your <b>side projects or other pers...
75,LanguageChoice,"Did you work with programming, scripting and m..."
76,Language,"Which <b>programming, scripting, and markup la..."


In [17]:
# Перевіряємо, чи є така колонка в df_p
'RemoteWork' in df_p.columns

True

In [18]:
# Дивимось, які значення є у колонці RemoteWork
df_p['RemoteWork'].value_counts(dropna=False)

RemoteWork
NaN                                                                             15411
Remote                                                                          10931
Hybrid (some remote, leans heavy to in-person)                                   6732
In-person                                                                        6042
Hybrid (some in-person, leans heavy to flexibility)                              5831
Your choice (very flexible, you can come in when you want or just as needed)     4244
Name: count, dtype: int64

In [19]:
# Кількість респондентів, які працюють віддалено
remote_workers_count = df_p[df_p['RemoteWork'] == 'Remote'].shape[0]
remote_workers_count

10931

ВИСНОВОК: 

10 931 респондент зазначив формат роботи Remote, що свідчить про значну поширеність повністю віддаленої роботи серед учасників опитування.

Завдання 5. Визначення популярності Python
-
- Інструмент, який потрібен для виконання цього завдання: Pandas (робота з текстовими даними, підрахунок відсотків)

- Опис завдання: Обчисли відсоток респондентів, які програмують на Python.

- Зверни увагу, що поле з мовами програмування може містити множинні значення, тому потрібно перевірити наявність Python в цьому полі.

In [20]:
# Шукаємо назву колонки, яка містить інформацію про мови програмування, якою програмують респонденти
[col for col in df_p.columns if 'Language' in col]

['LanguageChoice',
 'LanguageHaveWorkedWith',
 'LanguageWantToWorkWith',
 'LanguageAdmired',
 'LanguagesHaveEntry',
 'LanguagesWantEntry']

In [21]:
df_p ['LanguageHaveWorkedWith'].head()

0                     Bash/Shell (all shells);Dart;SQL
1                                                 Java
2                  Dart;HTML/CSS;JavaScript;TypeScript
3                                      Java;Kotlin;SQL
4    C;C#;C++;Delphi;HTML/CSS;Java;JavaScript;Lua;P...
Name: LanguageHaveWorkedWith, dtype: object

In [22]:
# Виявляємо респондентів, які працюють з Python
total_respondents = df_p.shape[0]
python_users = df_p['LanguageHaveWorkedWith'].str.contains('Python', case = False, na = False)

# Кількість Python-розробників
python_users_count = python_users.sum()

# Відсоток респондентів, які програмують на Python
python_percentage = (python_users_count/total_respondents) * 100
round(python_percentage, 1)

np.float64(37.5)

ВИСНОВОК:

Python використовують приблизно 37.5 % респондентів, що підтверджує його високу популярність серед учасників опитування.

Завдання 6. Аналіз шляхів навчання програмуванню
-
- Інструмент, який потрібен для виконання цього завдання: Pandas (фільтрація даних, робота з категоріальними змінними)

- Опис завдання:
> - Визнач, скільки респондентів навчалося програмувати за допомогою онлайн курсів.
> - Знайди відповідне поле в датасеті, яке описує способи навчання, та підрахуй тих, хто обрав онлайн курси.

In [23]:
# Знаходимо всі колонки, в назві яких є слово 'Learn'
[col for col in df_p.columns if 'Learn' in col]

['LearnCodeChoose', 'LearnCode', 'LearnCodeAI', 'AILearnHow']

In [24]:
# Перевіряємо, що саме зберігається у колонці LearnCode
df_p['LearnCode'].head()

0    Online Courses or Certification (includes all ...
1    Online Courses or Certification (includes all ...
2    Online Courses or Certification (includes all ...
3    Other online resources (e.g. standard search, ...
4                                                  NaN
Name: LearnCode, dtype: object

In [25]:
# Виявляємо респондентів, які навчалися за допомогою онлайн курсів
online_course_learners = df_p['LearnCode'].str.contains('Online Courses', case=False, na=False)

# Кількість таких респондентів
online_course_learners_count = online_course_learners.sum()
online_course_learners_count

np.int64(10973)

ВИСНОВОК: 

10 973 респонденти зазначили онлайн-курси як один зі способів навчання програмуванню, що підкреслює важливу роль онлайн-освіти у формуванні технічних навичок спеціалістів.

Завдання 7. Географічний аналіз компенсації Python-розробників
-
- Інструмент, який потрібен для виконання цього завдання: Pandas (групування даних, агрегація, фільтрація)

- Опис завдання: Серед респондентів, що програмують на Python, проведи аналіз компенсації (ConvertedCompYearly) в розрізі країн.

- Для кожної країни обчисли:
> - Середню суму компенсації
> - Медіанну суму компенсації

In [26]:
# Фільтрація респондентів, які програмують на Python
python_devs = df_p[df_p['LanguageHaveWorkedWith'].str.contains('Python', case = False, na = False)]

In [27]:
# 7.1 Відбір даних з компенсацією та країною

# Залишаємо лише потрібні колонки
python_devs_comp = python_devs[['Country', 'ConvertedCompYearly']]

# Видаляємо рядки з пропущеною компенсацією
python_devs_comp = python_devs_comp.dropna(subset=['ConvertedCompYearly'])

In [28]:
# 7.2 Групування по країнах та агрегація

country_compensation = (python_devs_comp
                        .groupby ('Country') ['ConvertedCompYearly']
                        .agg(
                            mean_compensation = 'mean',
                            median_compensation = 'median'
                        )
                        .reset_index()
                        .round()
                       )
country_compensation

,Country,mean_compensation,median_compensation
0,Afghanistan,22329.0,1000.0
1,Albania,47218.0,50000.0
2,Algeria,20187.0,7088.0
3,Andorra,226104.0,226104.0
4,Antigua and Barbuda,1.0,1.0
...,...,...,...
148,"Venezuela, Bolivarian Republic of...",9909.0,3000.0
149,Viet Nam,218837.0,8254.0
150,Yemen,32930.0,23672.0
151,Zambia,5424.0,3206.0


In [29]:
# 7.3 Сортуємо за середньою компенсацією (за спаданням)
country_compensation.sort_values(
    by='mean_compensation',
    ascending=False
).head(10)

,Country,mean_compensation,median_compensation
102,Oman,390135.0,390135.0
3,Andorra,226104.0,226104.0
149,Viet Nam,218837.0,8254.0
145,United States of America,173299.0,150000.0
132,Switzerland,156457.0,142592.0
121,Singapore,147515.0,91391.0
65,Israel,135828.0,142594.0
98,Nigeria,128773.0,1808.0
63,Ireland,120524.0,116015.0
99,Nomadic,120132.0,139218.0


ВИСНОВОК:

У межах завдання було проведено аналіз річної компенсації Python-розробників у розрізі країн. Дані було відфільтровано за використанням мови Python та наявністю значення компенсації. 
Для кожної країни обчислено середню та медіанну компенсацію. Результати показали значні відмінності між країнами, а також різницю між середнім і медіанним значеннями в окремих країнах (Viet Nam, Nigeria — дуже висока середня, але низька медіана), що може свідчити про вплив викидів або невелику кількість респондентів у вибірці.

Завдання 8. Аналіз освіти найбільш оплачуваних спеціалістів
-
- Інструмент, який потрібен для виконання цього завдання: Pandas (сортування даних, вибірка топ-записів)

- Опис завдання: Визнач рівні освіти 5 респондентів з найбільшою компенсацією. Тобі потрібно:
> - Відсортувати респондентів за рівнем компенсації
> - Вибрати топ-5
> - Показати їх рівні освіти

In [30]:
# Вибираємо потрібні колонки
comp_education = df_p[['ConvertedCompYearly', 'EdLevel']]

# Видаляємо респондентів без інформації про компенсацію
comp_education = comp_education.dropna(subset = ['ConvertedCompYearly'])

In [31]:
# Сортуємо за компенсацією у спадному порядку
top_5_paid = comp_education.sort_values (by = 'ConvertedCompYearly', ascending = False).head()
top_5_paid[['EdLevel', 'ConvertedCompYearly']]

,EdLevel,ConvertedCompYearly
34267,"Associate degree (A.A., A.S., etc.)",50000000.0
28700,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)",33552715.0
43143,"Associate degree (A.A., A.S., etc.)",18387548.0
35353,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)",15430267.0
45971,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)",13921760.0


ВИСНОВОК:

У межах завдання було проаналізовано рівні освіти п’яти респондентів з найвищою річною компенсацією. 
Дані були відсортовані за рівнем компенсації у спадному порядку, після чого визначено освітній рівень найбільш оплачуваних спеціалістів. 
Результати показали, що серед топ-5 представлені різні рівні освіти (Associate, Bachelor та Master), що свідчить про відсутність прямої залежності між рівнем формальної освіти та максимальною компенсацією.

*Завдання 9. Аналіз популярності Python по віковим категоріям 
-
*(виконання необов'язкове, за бажанням)

- Інструмент, який потрібен для виконання цього завдання: Pandas (групування, підрахунок відсотків, агрегація)

- Опис завдання: Для кожної вікової категорії визнач відсоток респондентів, які програмують на Python. Це дасть уявлення про те, в яких вікових групах Python є найпопулярнішим.

In [32]:
# Перевіряємо всі наявні вікові категорії
df_p['Age'].value_counts()

Age
25-34 years old      16519
35-44 years old      13241
18-24 years old       9210
45-54 years old       6275
55-64 years old       2626
65 years or older      942
Prefer not to say      378
Name: count, dtype: int64

In [33]:
# 9.1 Підготовка даних

# Залишаємо лише потрібні колонки
age_language = df_p [['Age','LanguageHaveWorkedWith']]

# Видаляємо респондентів без вікової категорії
age_language = age_language.dropna(subset = ['Age'])

In [34]:
# 9.2 Додаємо ознаку використання Python

# Створюємо булеву колонку, яка показує, чи використовує респондент Python
age_language['use_python'] = age_language['LanguageHaveWorkedWith'].str.contains('Python', case = False, na = False)

In [35]:
# 9.3 Групування по вікових категоріях

# Групуємо респондентів за віковими категоріями
python_by_age = (age_language
                 .groupby('Age')
                 .agg(
                     total_respondents = ('use_python', 'count'),
                     python_users = ('use_python', 'sum')
                 )
                )

# та рахуємо загальну кількість і кількість Python-розробників
python_by_age ['python_percentage'] = (
    (python_by_age['python_users']/python_by_age['total_respondents'])*100
).round()

python_by_age

,total_respondents,python_users,python_percentage
Age,,,
18-24 years old,9210,3684,40.0
25-34 years old,16519,6102,37.0
35-44 years old,13241,4862,37.0
45-54 years old,6275,2424,39.0
55-64 years old,2626,978,37.0
65 years or older,942,298,32.0
Prefer not to say,378,118,31.0


ВИСНОВОК: 

У межах завдання було проаналізовано популярність мови програмування Python серед різних вікових категорій респондентів. 
Результати показали, що частка Python-розробників є відносно стабільною у більшості вікових груп (близько 37–40%), з найвищими значеннями у групах 18–24 та 45–54 роки. У старших вікових категоріях (65 років і старше) частка Python-користувачів є нижчою, що може свідчити про більшу популярність Python серед молодших та середньовікових фахівців.

*Завдання 10. Аналіз індустрій серед високооплачуваних віддалених працівників
-
*(виконання необов'язкове, за бажанням)

- Інструмент, який потрібен для виконання цього завдання: Pandas (перцентилі, множинна фільтрація, підрахунок частоти)

- Опис завдання: Знайди респондентів, які:
> - Знаходяться у 75 перцентилі за компенсацією (тобто заробляють більше, ніж 75% респондентів)
> 
> - Працюють віддалено

Для цієї групи визнач найрозповсюдженіші індустрії.

In [36]:
# 10.1 Підготовка даних і 75-й перцентиль

# Видаляємо респондентів без інформації про компенсацію
comp_data = df_p.dropna (subset = 'ConvertedCompYearly')

# Обчислюємо 75-й перцентиль компенсації
percentile_75 = comp_data['ConvertedCompYearly'].quantile(0.75)
percentile_75

np.float64(120596.0)

In [37]:
# 10.2 Фільтрація: працівників 1) з компенсацією вище 75-го перцентиля; 2) які працюють у повністю віддаленому форматі

high_paid_remote = comp_data [
    (comp_data['ConvertedCompYearly'] > percentile_75)&
    (comp_data['RemoteWork'] == 'Remote')
]

In [38]:
# 10.3 Аналіз індустрій серед високооплачуваних працівників, які працюють віддалено

# Видаляємо респондентів без зазначеної індустрії
industry_data = high_paid_remote.dropna(subset=['Industry'])

# Підрахунок популярності індустрій
industry_counts = industry_data['Industry'].value_counts()

industry_counts

Industry
Software Development                          1186
Fintech                                        190
Healthcare                                     188
Other:                                         176
Internet, Telecomm or Information Services     138
Banking/Financial Services                      88
Government                                      78
Media & Advertising Services                    75
Retail and Consumer Services                    65
Transportation, or Supply Chain                 63
Computer Systems Design and Services            59
Insurance                                       46
Energy                                          45
Manufacturing                                   33
Higher Education                                30
Name: count, dtype: int64

In [39]:
# Перетворюємо у DataFrame
industry_table = industry_counts.reset_index()
industry_table.columns = ['Industry', 'Count']

industry_table


,Industry,Count
0,Software Development,1186
1,Fintech,190
2,Healthcare,188
3,Other:,176
4,"Internet, Telecomm or Information Services",138
5,Banking/Financial Services,88
6,Government,78
7,Media & Advertising Services,75
8,Retail and Consumer Services,65
9,"Transportation, or Supply Chain",63


ВИСНОВОК:

У межах завдання було проаналізовано індустрії серед респондентів, які входять до верхнього квартиля за рівнем річної компенсації та працюють у віддаленому форматі. 
Результати показали, що серед високооплачуваних віддалених працівників найбільш поширеною є індустрія Software Development, що значно випереджає інші сфери. Також серед популярних індустрій представлені "Fintech", "Healthcare" та "Internet, or Information Services", що свідчить про високий рівень цифровізації та попит на віддалену працю саме в цих галузях.

PS До аналізу також включено категорію «Prefer not to say», оскільки вона є валідним значенням у полі віку, а не пропущеним значенням.